### PR TITLE
Feature/full iac deployment

### PR DESCRIPTION
**Pull request by jan-zajac-dt**
# Description

Please provide a description of the changes you have made in this PR.

# How has this been tested?

Please explain how you have tested the new changes.

# Checklist

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have successfully run the E2E tests, and have included the links to the pipeline runs below
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have updated any relevant documentation to reflect my changes
- [ ] I have assigned a reviewer and messaged them

# Pipeline run links:

### CODE REVIEW

File: env.sh.example [hunk 1]
```diff
@@ -14,13 +14,17 @@
 # limitations under the License.
 
 export PAYLOAD=dev.json
-export PIPELINE_FILES_GCS_PATH=gs://<my-bucket>/pipelines
+export PIPELINE_FILES_BUCKET_NAME=jan-pipelines-bucket
```
Comment (caiharry) [inline env.sh.example:17]: These values need to be general

File: terraform/iam.tf [hunk 2]
```diff
@@ -21,8 +21,7 @@ module "pipeline_root_bucket_iam" {
   bucket_name = module.gcs_buckets["pipeline_root_bucket"].name
   member      = "serviceAccount:${module.service_accounts["pipelines_sa"].email}"
   roles = [
-    "roles/storage.objectViewer",
-    "roles/storage.objectCreator",
+    "roles/storage.admin" # Added by JAN, this trumps the reason above by giving even higher privileges as because get mysterious AttributeError. See post https://www.mail-archive.com/issues@beam.apache.org/msg183103.html. Example Pipeline: https://console.cloud.google.com/vertex-ai/locations/europe-west4/pipelines/runs/xgboost-train-pipeline-20220912123450?project=dt-jan-sandbox-dev
```
Comment (caiharry) [inline terraform/iam.tf:24]: No need to say who made this change, we can look in git logs for that. Keep the comment describing why we need `storage.admin`, though.

As this is a PR to the open source template, I'd recommend removing the link to your dt project because users outside dt won't have access to that

File: terraform/main.tf [hunk 3]
```diff
@@ -40,6 +40,7 @@ module "gcs_buckets" {
   version    = "~> 3.2"
   prefix     = ""
   project_id = var.project_id
+  location   = var.vertex_region # Added by JAN - This is missing my sandbox in comparison to the mlops playgrounds pipeline runner SA. I am hoping it will solve the issue with the model training job restarting because of error: https://console.cloud.google.com/logs/query;cursorTimestamp=2022-09-07T08:51:07.281723787Z;pinnedLogId=2022-09-07T08:40:56.206185934Z%2F1fweneyc9ca;query=severity%3DERROR%0Atimestamp%3D%222022-09-07T08:51:07.281723787Z%22%0AinsertId%3D%22pt460hc3cb%22;timeRange=2022-09-07T08:04:58.023Z%2F2022-09-07T11:04:58.023596Z?project=dt-jan-sandbox-dev
```
Comment (caiharry) [inline terraform/main.tf:43]: Before finalising this PR, remove reference to your sandbox project

File: terraform/main.tf [hunk 4]
```diff
@@ -102,3 +103,16 @@ resource "google_app_engine_application" "app" {
   location_id = var.app_engine_region
   depends_on  = [module.api_services]
 }
+
```
Comment (caiharry) [inline terraform/main.tf:106]: Maybe add a comment describing what these bq datasets are for?

File: terraform/terragrunt.hcl [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  cloud_schedulers_config = {
+    training = {
+      name         = "training-pipeline-trigger",
+      region       = "europe-west2",  # There are no cloud functions in europe-west4
+      description  = "Trigger my training pipeline in Vertex",
+      schedule     = "0 0 * * 0",
+      time_zone    = "UTC",
+      payload_file = "../pipelines/xgboost/training/payloads/dev.json",  # TODO: Think about this, the attributes.template_path references local file and not gcs location
```
Comment (caiharry) [inline terraform/terragrunt.hcl:79]: I think this is fine, we want to reference a local payload for quick iterations during development

You can use `get_terragrunt_dir()` or one of the other terragrunt functions to make the relative filepath more robust

File: terraform/terragrunt.hcl [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  ]
+
+  service_accounts = {
+    pipelines_sa = {
+      name         = get_env("VERTEX_SA_NAME"),
+      display_name = "Vertex Pipelines SA",
+      project_roles = [
+        "roles/aiplatform.user",
+        "roles/bigquery.jobUser",
+        "roles/bigquery.dataEditor", # Added by JAN - Getting 403 when trying to access taxi_trips table inside dataset: https://console.cloud.google.com/vertex-ai/locations/europe-west4/pipelines/runs/xgboost-train-pipeline-20220906174714?project=dt-jan-sandbox-dev AND https://stackoverflow.com/questions/52533796/unable-to-run-query-against-bigquery-permission-error-403
```
Comment (caiharry) [inline terraform/terragrunt.hcl:48]: As above, edit this comment

File: terraform/variables.tf [hunk 7]
```diff
@@ -35,6 +35,11 @@ variable "vertex_region" {
   type        = string
 }
 
+variable "bigquery_location" {
```
Comment (caiharry) [inline terraform/variables.tf:38]: How does this work with `transfer_datasets.sh`, for EU/non-EU locations? Pls update the docs (I think `README.md`) if they need updating

File: terraform/iam.tf (hunk 2 above)
Comment (jan-zajac-dt) [reply terraform/iam.tf:24]: 👍

File: terraform/main.tf (hunk 3 above)
Comment (jan-zajac-dt) [reply terraform/main.tf:43]: 👍

File: terraform/terragrunt.hcl (hunk 6 above)
Comment (jan-zajac-dt) [reply terraform/terragrunt.hcl:48]: 👍

File: terraform/variables.tf (hunk 7 above)
Comment (jan-zajac-dt) [reply terraform/variables.tf:38]: `transfer_datasets.sh` is used to copy data from a US region to EU so no need to update documentation.

 I've hardcoded in the `terragrunt.hcl` to be EU as I there are a ton of other places where location is a parameter that defaults to EU in the code.

File: cloudbuild/release.yaml [hunk 8]
```diff
@@ -43,4 +43,27 @@ steps:
         cp -r pipelines/${_PIPELINE_TEMPLATE}/prediction/assets ${TAG_NAME}/prediction/ && \
         gsutil cp -r ${TAG_NAME} ${_PIPELINE_PUBLISH_GCS_PATH}
 
-timeout: 1800s  # 30mins
+  # Install Python deps
```
Comment (browningjp-datatonic) [inline cloudbuild/release.yaml:46]: rebase

File: envs/terragrunt.hcl [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+      project_roles = [
+        "roles/aiplatform.user"
+      ],
+    },
+  }
+  gcs_buckets_names = {
+    pipeline_root_bucket      = "${local.project_id}-pipeline-root-bucket"
+    cf_staging_bucket         = "${local.project_id}-cf-staging-bucket"
+    compiled_pipelines_bucket = "${local.project_id}-compiled-pipelines-bucket"
+    assets_bucket             = "${local.project_id}-pipeline-assets-bucket"
```
Comment (browningjp-datatonic) [inline envs/terragrunt.hcl:70]: I think `compiled_pipelines_bucket` and `assets_bucket` are actually the same in the CI/CD, can be reduced to one bucket for both

File: envs/test/module/terragrunt.hcl [hunk 10]
```diff
@@ -0,0 +1,4 @@
+include "module" {
```
Comment (browningjp-datatonic) [inline envs/test/module/terragrunt.hcl:1]: Don't understand what these files are for in terragrunt!

**Review by browningjp-datatonic — CHANGES_REQUESTED** [review]
Really nice - please rebase to incorporate other changes

File: envs/terragrunt.hcl [hunk 11]
```diff
@@ -1,14 +1,21 @@
...
+}
+
+# Local values are overwritten by environment specific terragrunt config values
 locals {
-  project_id = get_env("VERTEX_PROJECT_ID")
-  region     = get_env("VERTEX_REGION")
+  env_config  = read_terragrunt_config(find_in_parent_folders("env.hcl"))
+  project_id  = local.env_config.locals.project_id
+  environment = local.env_config.locals.environment
+  region      = get_env("VERTEX_REGION")
```
Comment (efbbrown-dt) [inline envs/terragrunt.hcl:10]: This may just be my ignorance but would we also want to switch the way region is being declared the same as we've done for project_id?

File: Makefile [hunk 12]
```diff
@@ -28,10 +28,10 @@ trigger-tests: ## Runs unit tests for the pipeline trigger code
 	@pipenv run python -m pytest tests/trigger
 
 deploy-infra: ## Deploy infrastructure as defined in ./terraform directory using env variables from env.sh
```
[the remark this answers is not in the document]
Comment (efbbrown-dt) [reply Makefile:30]: On this point, don't think its necessary to have the last words as that is true for all commands.

```suggestion
deploy-infra: ## Deploy infrastructure as defined in ./terraform directory
```

File: Makefile [hunk 13]
```diff
@@ -27,6 +27,12 @@ unit-tests: ## Runs unit tests for kfp_components
 trigger-tests: ## Runs unit tests for the pipeline trigger code
 	@pipenv run python -m pytest tests/trigger
 
+deploy-infra: ## Deploy infrastructure as defined in ./terraform directory using env variables from env.sh
+	@terragrunt apply --terragrunt-working-dir="envs/dev/module"
+
+destroy-infra: ## Destroy infrastructure defined in ./terraform directory using env variables from env.sh
```
Comment (efbbrown-dt) [inline Makefile:33]: Same as above

```suggestion
destroy-infra: ## Destroy infrastructure defined in ./terraform directory
```

File: env.sh.example [hunk 14]
```diff
@@ -14,13 +14,18 @@
...
+
+export PIPELINE_FILES_BUCKET_NAME=${VERTEX_PROJECT_ID}-pipeline-assets-bucket
+export PIPELINE_ROOT_BUCKET_NAME=${VERTEX_PROJECT_ID}-pipeline-root-bucket
+export VERTEX_SA_NAME=vertex-pipelines
+
+export PIPELINE_FILES_GCS_PATH=gs://${PIPELINE_FILES_BUCKET_NAME}/pipelines
 # pipeline template - update to any pipelines under the pipelines folder
 # tensorflow or xgboost
-export PIPELINE_TEMPLATE=tensorflow
+export PIPELINE_TEMPLATE=xgboost
```
Comment (efbbrown-dt) [inline env.sh.example:26]: Seems like an unnecessary change

File: envs/prod/env.hcl [hunk 15]
```diff
@@ -0,0 +1,4 @@
+locals {
+  project_id  = "my-prod-project"
```
Comment (efbbrown-dt) [inline envs/prod/env.hcl:2]: Is defined as `get_env("VERTEX_PROJECT_ID")` in `envs/dev/env.hcl`, is there a reason its hard coded here?

Comment (browningjp-datatonic) [reply envs/prod/env.hcl:2]: I guess the idea is that if you are just wanting to try out the repo, spin up the environment by running terragrunt locally. If you want the full-fledged multi-project deployment, you would run this in CI/CD

Prod project would never be the same as the env variable set in env.sh for local development

File: terraform/iam.tf [hunk 16]
```diff
@@ -21,8 +21,7 @@ module "pipeline_root_bucket_iam" {
   bucket_name = module.gcs_buckets["pipeline_root_bucket"].name
   member      = "serviceAccount:${module.service_accounts["pipelines_sa"].email}"
   roles = [
-    "roles/storage.objectViewer",
-    "roles/storage.objectCreator",
+    "roles/storage.admin"
```
[the remark this answers is not in the document]
Comment (efbbrown-dt) [reply terraform/iam.tf:24]: I think the best combo of roles here is:

`roles/storage.legacyBucketWriter` and `roles/storarage.objectAdmin`. Without the legacy bucket writer the pipeline-sa won't be able to do anything inside the bucket, but it removes the controversial perms associated with `storage.admin`.

File: terraform/terragrunt.hcl [hunk 17]
```diff
@@ -57,10 +57,10 @@ inputs = {
     },
   }
   gcs_buckets_names = {
-    pipeline_root_bucket      = get_env("PIPELINE_ROOT_BUCKET_NAME")
+    pipeline_root_bucket      = "${local.project_id}-pipeline-root-bucket"
```
Comment (jan-zajac-dt) [inline terraform/terragrunt.hcl]: Why are we switching from get_env? Means we have bucket names in two places. Same fir assets bucket below

File: terraform/iam.tf (hunk 16 above)
[the remark this answers is not in the document]
Comment (browningjp-datatonic) [reply terraform/iam.tf:24]: Don't think you need `roles/storage.legacyBucketWriter` if you have `roles/storage.objectAdmin`
